extensions/typescript-language-features/src/languageFeatures/diagnostics.ts

```diff
@@ -148,6 +151,76 @@ class DiagnosticSettings {
 	}
 }
 
+class DiagnosticsTelemetryManager extends Disposable {
+
+	private readonly _diagnosticCodesMap = new Map<number, number>();
+	private readonly _diagnosticSnapshotsMap = new Map<string, readonly vscode.Diagnostic[]>();
+	private _timeout: NodeJS.Timeout | undefined;
+
+	constructor(
+		private readonly _telemetryReporter: TelemetryReporter,
+		private readonly _diagnosticsCollection: vscode.DiagnosticCollection,
+	) {
+		super();
+		this._register(vscode.workspace.onDidChangeTextDocument(e => {
+			if (e.document.languageId === 'typescript') {
+				this._updateAllDiagnosticCodesAfterTimeout();
+			}
+		}));
+		this._updateAllDiagnosticCodesAfterTimeout();
+		this._registerTelemetryEventEmitter();
+	}
+
+	private _updateAllDiagnosticCodesAfterTimeout() {
+		clearTimeout(this._timeout);
+		this._timeout = setTimeout(() => this._updateDiagnosticCodes(), 5000);
+	}
+
+	private _increaseDiagnosticCodeCount(code: string | number | undefined) {
+		if (code === undefined) {
+			return;
+		}
+		this._diagnosticCodesMap.set(Number(code), (this._diagnosticCodesMap.get(Number(code)) || 0) + 1);
+	}
+
+	private _updateDiagnosticCodes() {
+		this._diagnosticsCollection.forEach((uri, diagnostics) => {
+			const uriString = uri.toString();
+			const previousDiagnostics = this._diagnosticSnapshotsMap.get(uriString);
+			this._diagnosticSnapshotsMap.set(uriString, diagnostics);
+			const diagnosticsDiff = diagnostics.filter((diagnostic) => !previousDiagnostics?.some((previousDiagnostic) => equals(diagnostic, previousDiagnostic)));
```

Comment:
I believe this will also update the diagnostics counts when they shift around as you type (since the objects would no longer be equal). Is there some way to avoid this as I imagine it's not helpful

Reply:
Hey @mjbvz, thanks for reviewing the PR. To mitigate this, we update the diagnostics after 5 seconds of no typing. In this case we consider it to be a pause in typing which requires recomputation of the diagnostics. Perhaps we could increase the timeout to a larger value in case the user stops to think. When the user continously types, there is no diagnostic recalculation.